File: core/ipbin_editor.py
"""
core/ipbin_editor.py — Dreamcast IP.BIN reader and editor
M5 Dev | GPL v3

IP.BIN is the Dreamcast boot sector: 2048 bytes, fixed ASCII/Shift-JIS fields.
Can be read from a standalone IP.BIN file or extracted from a GDI/CHD.
"""

import os

IPBIN_SIZE = 2048

# Field definitions: (offset, length, name)
IPBIN_FIELDS = [
    (0x00, 16, 'hardware_id'),      # "SEGA SEGADREAMCAST "
    (0x10, 16, 'maker_id'),         # "SEGA ENTERPRISES"
    (0x20, 10, 'product_number'),   # e.g. "T-0000"
    (0x2A,  6, 'product_version'),  # e.g. "V1.001"
    (0x30,  8, 'release_date'),     # "YYYYMMDD"
    (0x38,  8, 'boot_filename'),    # e.g. "1ST_READ.BIN"
    (0x40, 16, 'software_type'),    # "  GD-ROM1/1     "
    (0x50,  8, 'region_code'),      # "JUE     " (J=Japan, U=USA, E=Europe)
    (0x58,  8, 'peripherals'),      # bitfield as ASCII hex
    (0x60, 16, 'product_name'),     # game title
    (0x70, 16, 'product_name_2'),   # game title continued
]

REGION_FLAGS = {'J': 'Japan', 'U': 'USA', 'E': 'Europe'}


def read_ipbin(path: str) -> dict:
    """
    Read an IP.BIN file and return a dict of its fields.
    Raises ValueError if file is not a valid Dreamcast IP.BIN.
    """
    if not os.path.isfile(path):
        raise FileNotFoundError(f'IP.BIN not found: {path}')

    with open(path, 'rb') as f:
        data = f.read(IPBIN_SIZE)

    if len(data) < IPBIN_SIZE:
        raise ValueError(f'File too small to be IP.BIN: {len(data)} bytes (need {IPBIN_SIZE})')

    # Validate Dreamcast magic
    hardware_id = data[0x00:0x10].decode('ascii', errors='replace').strip()
    if 'SEGA' not in hardware_id:
        raise ValueError(f'Not a valid Dreamcast IP.BIN (hardware_id: {hardware_id!r})')

    result = {'_raw': data, '_path': path}
    for offset, length, name in IPBIN_FIELDS:
        raw = data[offset: offset + length]
        if name in ('product_name', 'product_name_2'):
            try:
                decoded = raw.decode('shift_jis')
            except UnicodeDecodeError:
                try:
                    decoded = raw.decode('cp932')
                except UnicodeDecodeError:
                    decoded = raw.decode('ascii', errors='replace')
        else:
            decoded = raw.decode('ascii', errors='replace')
        result[name] = decoded.rstrip()

    # Parse region flags
    region_raw = result.get('region_code', '')
    result['regions'] = [REGION_FLAGS[c] for c in 'JUE' if c in region_raw]

    return result


def write_ipbin(fields: dict, output_path: str):
    """
    Write modified IP.BIN fields back to a file.
    fields must contain '_raw' (the original 2048-byte blob).
    Only fields present in IPBIN_FIELDS are updated.
    """
    data = bytearray(fields['_raw'])

    for offset, length, name in IPBIN_FIELDS:
        if name not in fields or name.startswith('_'):
            continue
        val = fields[name]
        try:
            encoded = val.encode('ascii')
        except UnicodeEncodeError:
            try:
                encoded = val.encode('shift_jis')
            except UnicodeEncodeError:
                encoded = val.encode('ascii', errors='replace')

        # Pad with spaces to field length, truncate if too long
        padded = encoded[:length].ljust(length, b' ')
        data[offset: offset + length] = padded

    # Update region code from regions list
    if 'regions' in fields:
        region_str = ''
        for code, name in REGION_FLAGS.items():
            if name in fields['regions']:
                region_str += code
        region_encoded = region_str.encode('ascii').ljust(8, b' ')
        data[0x50:0x58] = region_encoded

    with open(output_path, 'wb') as f:
        f.write(bytes(data))


def set_region_free(fields: dict) -> dict:
    """Enable all three regions (Japan, USA, Europe)."""
    fields['regions'] = ['Japan', 'USA', 'Europe']
    return fields


def set_vga_enabled(fields: dict) -> dict:
    """
    Enable VGA output in the peripheral flags byte.
    Bit 4 of the peripherals field (offset 0x58) controls VGA.
    """
    perip = fields.get('peripherals', '00000000')
    try:
        flags = int(perip.strip(), 16)
    except ValueError:
        flags = 0
    flags |= (1 << 4)  # bit 4 = VGA box support
    fields['peripherals'] = f'{flags:08X}'
    return fields

File: core/lib/src/ipbin_editor.dart
// OpenROM — Universal ROM Compression Suite
// M5 Dev | GPL v3

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

const int ipbinSize = 2048;

class IpbinField {
  final int offset;
  final int length;
  final String name;

  const IpbinField(this.offset, this.length, this.name);
}

const List<IpbinField> ipbinFields = [
  IpbinField(0x00, 16, 'hardware_id'),
  IpbinField(0x10, 16, 'maker_id'),
  IpbinField(0x20, 10, 'product_number'),
  IpbinField(0x2A, 6, 'product_version'),
  IpbinField(0x30, 8, 'release_date'),
  IpbinField(0x38, 8, 'boot_filename'),
  IpbinField(0x40, 16, 'software_type'),
  IpbinField(0x50, 8, 'region_code'),
  IpbinField(0x58, 8, 'peripherals'),
  IpbinField(0x60, 16, 'product_name'),
  IpbinField(0x70, 16, 'product_name_2'),
];

const Map<String, String> regionFlags = {
  'J': 'Japan',
  'U': 'USA',
  'E': 'Europe',
};

Map<String, dynamic> readIpbin(String path) {
  final file = File(path);
  if (!file.existsSync()) {
    throw FileSystemException('IP.BIN not found: $path', path);
  }

  final data = file.readAsBytesSync();
  if (data.length < ipbinSize) {
    throw FormatException(
        'File too small to be IP.BIN: ${data.length} bytes (need $ipbinSize)');
  }

  final hardwareIdBytes = data.sublist(0x00, 0x10);
  final hardwareId = ascii.decode(hardwareIdBytes, allowInvalid: true).trim();
  if (!hardwareId.contains('SEGA')) {
    throw FormatException('Not a valid Dreamcast IP.BIN (hardware_id: "$hardwareId")');
  }

  final result = <String, dynamic>{
    '_raw': Uint8List.fromList(data.sublist(0, ipbinSize)),
    '_path': path,
  };

  for (final field in ipbinFields) {
    final raw = data.sublist(field.offset, field.offset + field.length);
    String decoded;
    if (field.name == 'product_name' || field.name == 'product_name_2') {
      try {
        decoded = latin1.decode(raw);
      } catch (_) {
        decoded = ascii.decode(raw, allowInvalid: true);
      }
    } else {
      decoded = ascii.decode(raw, allowInvalid: true);
    }
    result[field.name] = decoded.trimRight();
  }

  final regionRaw = result['region_code'] as String? ?? '';
  final regions = <String>[];
  for (final c in ['J', 'U', 'E']) {
    if (regionRaw.contains(c)) {
      regions.add(regionFlags[c]!);
    }
  }
  result['regions'] = regions;

  return result;
}

void writeIpbin(Map<String, dynamic> fields, String outputPath) {
  final raw = fields['_raw'] as Uint8List?;
  if (raw == null || raw.length < ipbinSize) {
    throw ArgumentError('fields must contain valid _raw Uint8List buffer of 2048 bytes');
  }

  final data = Uint8List.fromList(raw);

  for (final field in ipbinFields) {
    if (!fields.containsKey(field.name) || field.name.startsWith('_')) {
      continue;
    }
    final val = fields[field.name] as String? ?? '';
    List<int> encoded;
    try {
      encoded = ascii.encode(val);
    } catch (_) {
      encoded = latin1.encode(val);
    }

    final padded = List<int>.filled(field.length, 0x20);
    for (int i = 0; i < field.length && i < encoded.length; i++) {
      padded[i] = encoded[i];
    }

    for (int i = 0; i < field.length; i++) {
      data[field.offset + i] = padded[i];
    }
  }

  if (fields.containsKey('regions')) {
    final regionsList = (fields['regions'] as List).cast<String>();
    var regionStr = '';
    regionFlags.forEach((code, name) {
      if (regionsList.contains(name)) {
        regionStr += code;
      }
    });

    final regionEncoded = ascii.encode(regionStr);
    final padded = List<int>.filled(8, 0x20);
    for (int i = 0; i < 8 && i < regionEncoded.length; i++) {
      padded[i] = regionEncoded[i];
    }
    for (int i = 0; i < 8; i++) {
      data[0x50 + i] = padded[i];
    }
  }

  File(outputPath).writeAsBytesSync(data);
}

Map<String, dynamic> setRegionFree(Map<String, dynamic> fields) {
  fields['regions'] = ['Japan', 'USA', 'Europe'];
  return fields;
}

Map<String, dynamic> setVgaEnabled(Map<String, dynamic> fields) {
  final perip = fields['peripherals'] as String? ?? '00000000';
  int flags = 0;
  try {
    flags = int.parse(perip.trim(), radix: 16);
  } catch (_) {
    flags = 0;
  }
  flags |= (1 << 4);
  fields['peripherals'] = flags.toRadixString(16).toUpperCase().padLeft(8, '0');
  return fields;
}
